internal/moduletest/graph/eval_context.go

```diff
@@ -565,43 +568,80 @@ func diagsForEphemeralResources(refs []*addrs.Reference) (diags tfdiags.Diagnost
 	return diags
 }
 
-// UpdateStateFile updates the internal state file for a given state_key value.
-// The TestFileState argument is used to update only the exported fields in the
-// preexisting TestFileState value. Unexported fields' values are preserved
-// from when they are first set while building the test graph.
-//
-// If there isn't a state file for the given state_key then UpdateStateFile will use
-// the TestFileState argument to set its value.
-func (ec *EvalContext) UpdateStateFile(key string, state *TestFileState) {
+func (ec *EvalContext) SetFileState(key string, run *moduletest.Run, state *states.State, reason teststates.StateReason) {
 	ec.stateLock.Lock()
 	defer ec.stateLock.Unlock()
-	oldState, exists := ec.FileStates[key]
-
-	if !exists {
-		ec.FileStates[key] = state
-	}
 
-	ec.FileStates[key] = &TestFileState{
-		File:   state.File,
-		Run:    state.Run,
-		State:  state.State,
-		Reason: state.Reason,
-
-		backend: oldState.backend,
+	current := ec.FileStates[key]
+
+	// Whatever happens we're going to record the latest state for this key.
+	current.State = state
+	current.Manifest.Reason = reason
+
+	if run.Config.SkipCleanup {
+		// if skip cleanup is set on the run block, we're going to track it
+		// as the thing to target regardless of what else might be true.
+		current.Run = run
+
+		// we'll mark the state as being restored to the current run block
+		// if (a) we're not in cleanup mode (meaning everything should be
+		// destroyed) or (b) we are in cleanup mode but with the repair flag
+		// which means that only errored states should be destroyed.
+		current.RestoreState = ec.mode != moduletest.CleanupMode || ec.repair
+	} else if !current.RestoreState {
+		// otherwise, only set the new run block if we haven't been told the
+		// earlier run block is more relevant.
+		current.Run = run
 	}
 }
 
-func (ec *EvalContext) GetFileState(key string) *TestFileState {
+// GetState retrieves the current state for the specified key, exactly as it
+// specified within the current cache.
+func (ec *EvalContext) GetState(key string) *teststates.TestRunState {
 	ec.stateLock.Lock()
 	defer ec.stateLock.Unlock()
-	return ec.FileStates[key]
+	current := ec.FileStates[key]
+	if current == nil {
+		// this shouldn't happen, panic here so we now exactly where the origin
```

Comment:
Could the `SetFileState` method use this one to retrieve the current state instead, so it also benefits from the annotated panic message.